### PR TITLE
McAfeeNSM: Added new integration

### PR DESCRIPTION
**Pull request by yaakovi**
fixes: https://github.com/demisto/server/issues/5371

### CODE REVIEW

File: Integrations/integration-McAfeeNSM.yml [hunk 1]
```diff
@@ -0,0 +1,679 @@
...
+        // store current credentials
+        session_credentials = Base64.encode(res.session + ':' + res.userId);
+
+        return res;
+    }
+
+    function get_attacks(attack_id) {
+        var cmd_url, res;
+        if (attack_id) {
+            if (attack_id.match('^0x[0-9A-Fa-f]{8}$') === null) {
```
Comment (asafshen) [inline Integrations/integration-McAfeeNSM.yml]: nice

File: Integrations/integration-McAfeeNSM.yml [hunk 2]
```diff
@@ -0,0 +1,679 @@
...
+                return 'Error! Attack ID must be formated as 32-bit hexadecimal number. for example: 0x1234BEEF';
+            }
+
+            cmd_url = 'attack/{0}'.format(attack_id);
```
Comment (asafshen) [inline Integrations/integration-McAfeeNSM.yml]: I would use native JS (i.e. `'attack/' +  attack_id`). "ES5" way to do so.

Using format is also ok of course :)

File: Integrations/integration-McAfeeNSM.yml [hunk 3]
```diff
@@ -0,0 +1,679 @@
...
+
+        if (state) {
+            cmd_args.alertstate = state;
+        }
+
+        if (time_period) {
+            cmd_args.timeperiod = time_period;
+            if (time_period === 'CUSTOM') {
+                if ((start_time === undefined) || (end_time === undefined)) {
+                    return 'Error! In CUSTOM mode, You must specify both start time and end time';
```
Comment (asafshen) [inline Integrations/integration-McAfeeNSM.yml]: you don't do anything this error.
recommended options:
1. throw error
2. return empty string (`''`) and validate in caller 
3. don't validate this here, just write something like `function validateTimeArgs(time_period, start_time, end_time) {..}` and call it in the relevant places in the `switch` in the bottom of the script.

File: Integrations/integration-McAfeeNSM.yml [hunk 4]
```diff
@@ -0,0 +1,679 @@
...
+            cmd_args.filter = filter;
+        }
+
+        return cmd_url + encodeToURLQuery(cmd_args);
+    }
+
+    function get_alerts(state, time_period, start_time, end_time, search, filter) {
+        var cmd_url = all_alerts_url(state, time_period, start_time, end_time, search, filter);
+        var res = request('GET', cmd_url, {}, '');
+        log('Showing {0}/{1} Alerts:'.format(res.retrievedAlertsCount, res.totalAlertsCount));
```
Comment (asafshen) [inline Integrations/integration-McAfeeNSM.yml]: remove log plz

File: Integrations/integration-McAfeeNSM.yml [hunk 5]
```diff
@@ -0,0 +1,679 @@
...
+    }
+
+    function update_alerts(state, time_period, start_time, end_time, search, filter, new_state, new_assignee) {
+        var cmd_url = all_alerts_url(state, time_period, start_time, end_time, search, filter);
+        if ((new_state === undefined) && (new_assignee === undefined)) {
+            return 'Error! You must specify a new alert state or a new assignee';
+        }
+
+        var query = {};
+        if (new_state !== undefined) {
```
Comment (asafshen) [inline Integrations/integration-McAfeeNSM.yml]: in JS is more common to write  `if (!new_state)`

That's ok also of course :)

File: Integrations/integration-McAfeeNSM.yml [hunk 6]
```diff
@@ -0,0 +1,679 @@
...
+        return alert_to_entry('Alert: {0}'.format(res.name), res)
+    }
+
+    function get_ntba_monitors() {
+        var cmd_url = 'ntbamonitors';
+        var res = request('GET', cmd_url, {}, '');
+
+        return res;
+    }
+    function get_events(nba_id, hash, duration) {
```
Comment (asafshen) [inline Integrations/integration-McAfeeNSM.yml]: add new line plz

File: Integrations/integration-McAfeeNSM.yml [hunk 7]
```diff
@@ -0,0 +1,679 @@
...
+        var cmd_url = '{0}/endpointintelligence/{1}/events?duration={2}'.format(nba_id, hash, duration);
+        var res = request('GET', cmd_url, {}, '');
+
+        return res.eventList;
+    }
+
+    login();
+    // The command input arg holds the command sent from the user.
```
Comment (asafshen) [inline Integrations/integration-McAfeeNSM.yml]: remove this comment plz

File: Integrations/integration-McAfeeNSM.yml [hunk 8]
```diff
@@ -0,0 +1,679 @@
...
+        case 'nsm-get-attacks':
+            return get_attacks(args.attack_id);
+
+        case 'nsm-get-ips-policies':
+            return get_IPS_policies(args.domain_id);
+        case 'nsm-get-ips-policy-details':
+            return get_IPS_policy_details(args.policy_id);
+
+
+        // This is the call made when pressing the integration test button.
```
Comment (asafshen) [inline Integrations/integration-McAfeeNSM.yml]: remove this comment plz

File: Integrations/integration-McAfeeNSM.yml [hunk 9]
```diff
@@ -0,0 +1,679 @@
...
+        case 'test-module':
+            return 'ok';
```
Comment (asafshen) [inline Integrations/integration-McAfeeNSM.yml]: the test does not do anything and it should

usually, we call some general command that supposes to work always... maybe `get_domains()`?

File: Integrations/integration-McAfeeNSM.yml [hunk 10]
```diff
@@ -0,0 +1,679 @@
...
+        case 'fetch-incidents':
```
Comment (asafshen) [inline Integrations/integration-McAfeeNSM.yml]: please remove all script code below this line if it unused

File: Integrations/integration-McAfeeNSM.yml [hunk 11]
```diff
@@ -0,0 +1,679 @@
...
+    var insecure = params.insecure;
+    var base_url = params.url;
+    // handle '/' at the end of the url
+    if (base_url[base_url.length - 1] === '/') {
+        base_url += "sdkapi/";
+    } else {
+        base_url += "/sdkapi/";
+    }
+
+    function request(method, url_suffix, headers, body) {
```
Comment (asafshen) [inline Integrations/integration-McAfeeNSM.yml]: I recommend changing the name `sendRequest(..)`, that kinda convention I think...

File: TestPlaybooks/playbook-McAfeeNSMTest.yml [hunk 12]
```diff
@@ -0,0 +1,308 @@
...
+      "dimensions": {
+        "height": 1815,
+        "width": 380,
+        "x": 50,
+        "y": 50
+      }
+    }
+  }
+inputs: []
+releaseNotes: Added test to McAfee NSM integration
```
Comment (asafshen) [inline TestPlaybooks/playbook-McAfeeNSMTest.yml]: Not critical... but release notes are not necessary for test playbooks

File: Integrations/integration-McAfeeNSM.yml [hunk 13]
```diff
@@ -0,0 +1,679 @@
...
+    - contextPath: NSM.Attacks.Name
+      description: Attack Name
+    - contextPath: NSM.Attacks.Severity
+      description: Attack Severity
+    - contextPath: NSM.Attacks.Direction
+      description: Attack Direction
+    - contextPath: NSM.Attacks.Category
+      description: Attack Category
+    description: Get all available attack definitions in NSM
+releaseNotes: Add new integration, supporting alerts querying and updating, IPS policies querying
```
Comment (asafshen) [inline Integrations/integration-McAfeeNSM.yml]: Not critical but `releaseNotes` are not necessary for new integrations/scripts/etc

File: Integrations/integration-McAfeeNSM.yml [hunk 14]
```diff
@@ -0,0 +1,679 @@
...
+        return res;
+    };
+  type: javascript
+  commands:
+  - name: nsm-get-pcap
+    arguments:
+    - name: sensorID
+      required: true
+      default: true
+      description: the sensor ID related to the pcap
```
Comment (asafshen) [inline Integrations/integration-McAfeeNSM.yml]: Capital `T` (just to be consistent..)

File: Integrations/integration-McAfeeNSM.yml [hunk 15]
```diff
@@ -0,0 +1,679 @@
...
+    description: Get a pcap file of given event
+  - name: nsm-get-sensors
+    arguments:
+    - name: domainID
+      default: true
+      description: the domain of the sensors. If blank, returns all sensors.
```
Comment (asafshen) [inline Integrations/integration-McAfeeNSM.yml]: Capital `T` (just to be consistent..)

File: Integrations/integration-McAfeeNSM.yml [hunk 16]
```diff
@@ -0,0 +1,679 @@
...
+    - name: state
+      auto: PREDEFINED
+      predefined:
+      - ANY
+      - Acknowledged
+      - Unacknowledged
+      description: Alert State
+      defaultValue: ANY
+    - name: search
+      description: Search
```
Comment (asafshen) [inline Integrations/integration-McAfeeNSM.yml]: Supply example if you can

File: Integrations/integration-McAfeeNSM.yml [hunk 17]
```diff
@@ -0,0 +1,679 @@
...
+    - name: new_state
+      auto: PREDEFINED
+      predefined:
+      - Acknowledged
+      - Unacknowledged
+      description: The new alert state
+    - name: new_assignee
+      description: The new assignee
+    - name: search
+      description: Search
```
Comment (asafshen) [inline Integrations/integration-McAfeeNSM.yml]: Supply example if you can

File: Integrations/integration-McAfeeNSM.yml [hunk 18]
```diff
@@ -0,0 +1,679 @@
...
+    - contextPath: NSM.Alerts.Attack
+      description: Alert's Attack
+    - contextPath: NSM.Alerts.Attacker
+      description: The Attacker who committed the attack
+    - contextPath: NSM.Alerts.Target
+      description: The Attack's Target
+    - contextPath: NSM.Alerts.MalwareFile
+      description: Malware File used in the attack
+    description: Get alerts
+  - name: nsm-update-alerts
```
Comment (asafshen) [inline Integrations/integration-McAfeeNSM.yml]: Which alert does this command update?
What are the semantic of the alert state argument and the search argument?

File: Integrations/integration-McAfeeNSM.yml [hunk 19]
```diff
@@ -0,0 +1,679 @@
...
+    var base_url = params.url;
+    // handle '/' at the end of the url
+    if (base_url[base_url.length - 1] === '/') {
+        base_url += "sdkapi/";
+    } else {
+        base_url += "/sdkapi/";
+    }
+
+    function request(method, url_suffix, headers, body) {
+        logInfo('url: ' + url_suffix);
```
[the remark this answers is not in the document]
Comment (Spivaka) [reply Integrations/integration-McAfeeNSM.yml]: or supply a better message than url: ...

File: Integrations/integration-McAfeeNSM.yml (hunk 2 above)
Comment (Spivaka) [reply Integrations/integration-McAfeeNSM.yml]: Its always amazing that you can see which programming language is the no. 1 choice by a programmer code :)
but this is JS and not python :)

File: Integrations/integration-McAfeeNSM.yml (hunk 5 above)
Comment (Spivaka) [reply Integrations/integration-McAfeeNSM.yml]: please do change

File: Integrations/integration-McAfeeNSM.yml [hunk 20]
```diff
@@ -0,0 +1,679 @@
...
+  defaultvalue: ""
+  type: 8
+  required: false
+script:
+  script: |
+    var session_credentials = '';
+    var insecure = params.insecure;
+    var base_url = params.url;
+    // handle '/' at the end of the url
+    if (base_url[base_url.length - 1] === '/') {
```
Comment (Spivaka) [inline Integrations/integration-McAfeeNSM.yml]: you can have more than on... I recommend to write a code that remove them all and add /sdkapi

File: Integrations/integration-McAfeeNSM.yml [hunk 21]
```diff
@@ -0,0 +1,679 @@
...
+    }
+
+    function login() {
+        var cmd_url = "session";
+        var credentials = Base64.encode(params.credentials.identifier + ':' + params.credentials.password);
+        var headers = {'NSM-SDK-API' : [credentials]};
+        var res = request("GET", cmd_url, headers, '');
+
+        // store current credentials
+        session_credentials = Base64.encode(res.session + ':' + res.userId);
```
Comment (Spivaka) [inline Integrations/integration-McAfeeNSM.yml]: please talk to @yuvalbenshalom who added a base 64 using go (atob i think)

File: Integrations/integration-McAfeeNSM.yml [hunk 22]
```diff
@@ -0,0 +1,679 @@
...
+
+        return res;
+    }
+
+    function get_attacks(attack_id) {
+        var cmd_url, res;
+        if (attack_id) {
+            if (attack_id.match('^0x[0-9A-Fa-f]{8}$') === null) {
+                return 'Error! Attack ID must be formated as 32-bit hexadecimal number. for example: 0x1234BEEF';
```
Comment (Spivaka) [inline Integrations/integration-McAfeeNSM.yml]: why return and not throw?

File: Integrations/integration-McAfeeNSM.yml (hunk 9 above)
Comment (Spivaka) [reply Integrations/integration-McAfeeNSM.yml]: I think this is ok, as it just checks login (which is the most basic action)

Comment (yaakovi) [reply Integrations/integration-McAfeeNSM.yml]: in this case, the login is tested, it is executing before the switch.
i can add the get domain as well

**Pull request merged**